### DISCUSSION TITLE
Can xterm be used to connect devices via tcp instead of WebSocket?

### DISCUSSION THREAD

**Discussion by dbssAlan** [sourcelair/xterm.js#4751] [Q&A] [answered]
hi,I'm investigating a device log display scenario that involves scrolling through the interface to view logs, download log files, enter commands, and so on. But this is an `IoT` device, and it is connected to the outside via `TCP`. Our client is based on `Electron`, and I want to use `xterm` to display log messages sent by the device. But I see that `xterm` usually uses `WebSocket`, so I'm not sure if that's possible

**Comment by jerch** [comment, marked as answer]
It depends on how you set up your electron client:

JS in a browser context normally has no access to raw TCP sockets, it needs some sort of HTTP wrapping (either by Ajax/long polling or via websockets). Thats the reason why its often done with websockets.

But with electron the situation changes - you can either use the electron client process as proxy to translate between TCP and the browser context (again forward data with websockets or with IPC), or allow the browser context to load nodejs modules and hook TCP handling in there directly (via the `net` module).

In the end this is more an electron question than xterm.js related. To get data into xterm.js you only have to feed it to `term.write` - how you get the data to that point, is up to you.

**Reply by dbssAlan** [reply]
Yes, I found this point too, thank you very much! 😊 I'm going to write a `net` module-based plugin using the `xterm-addon-attach` plugin
